demo/repo/docker-test.sh

```diff
@@ -0,0 +1,17 @@
+set -ex
+cd `dirname $0`
+
+APP_IMG=localhost/examplecorp/spring-petclinic:$BUILD_TAG
+TEST_IMG=examplecorp/spring-petclinic-tests:$BUILD_TAG
+TEST_CTR=spring-petclinic-tests-$BUILD_TAG
+
+# Create the test container.
+docker build -t $TEST_IMG test
+
+# Run the application in a separate container.
+cid=`docker run -d $APP_IMG`
+trap "docker rm -f $cid $TEST_CTR" EXIT
+
+# Run the test container.
+docker run --name=$TEST_CTR --link=$cid:petclinic -v m2repo:/m2repo $TEST_IMG /src/test.sh
+docker cp $TEST_CTR:/src/screenshot.jpg .
```

Comment:
This part is a bit awkward—you would like `archiveArtifacts` to be able to grab the file straight out of the container. Theoretically there could be a wrapper step which sets a contextual `FilePath` “workspace” to the container’s `/` (or some other directory of your choice), allowing nested steps to transparently operate in the container’s filesystem.

Unfortunately the current design of `FilePath` makes this impossible: it is not an extensible VFS, but rather a `final` class with two implementations basically selected by a switch: local `File`, or remote `File` using `Channel` calls. There is no place for a plugin to create a completely alternate implementation. And you cannot create a `Channel` into the container without running `slave.jar` inside it, which would require it have some JRE installed.

The Kubernetes plugin support for Pipeline, and IIRC the `docker-slaves` plugin, try to work around this by running a little pod of containers using shared volumes, and running `slave.jar` inside a designated one. That is a bit heavier weight, though, since it requires running an extra JVM process, and you have to be prepared to designate interesting parts of your “real” container as volumes.

Another trick would be to implement `VirtualChannel` from scratch to handle selected `FilePath` callables using completely different implementations, but this would be extremely fragile without a lot of work in core.

Reply:
This is now easier using `--output` in Buildkit.